site/_posts/2018-02-14-the-pathname-to-discovering-a-memory-leak-in-ruby.html.md

```diff
@@ -0,0 +1,429 @@
+with no leak, but ended up trying it again.  The first draft of the script
+looked basically like this:
+
+```ruby
+# Code needed to get MiqProcess.processInfo loaded and working
+require 'manageiq-gems-pending'
+require 'util/miq-process'
+require 'active_support/all'
+
+pid = ARGV[0]
+
+loop do
+  MiqProcess.processInfo(pid)
+  GC.start
+  sleep 0.5
+end
+```
+
+Unfortunately, it didn't leak...
+
+┻━┻ ︵ヽ(\`Д´)ﾉ︵ ┻━┻
+
+...
+
+┬──┬ ノ(ò\_óノ)
+
+Well crap...
+
+That is when Keenan Brock (who I was pairing with at the time) had a brilliant
+idea, where he suggested that we include a call to `require
+config/environment.rb` to load and initialize our Rails app, and then run the
+script again.  What do you know, it leaked!  And on a regular basis as well!
+
+From there, we narrowed down what the core components of `.processInfo` were,
+and ran each bit overnight and logged the `[heap]` section of the
+`/proc/<PID>/smaps` overtime.  The leaky section ended up being a call to
+`MiqSystem.num_cpus`, and the code for that method can be seen below:
+
+```ruby
+def self.num_cpus
+  return unless Sys::Platform::IMPL == :linux
+  require 'linux_admin'
+  @num_cpus ||= LinuxAdmin::Hardware.new.total_cores
+end
+```
+
+Looks completely harmless... what gives!
+
+Well, this method does a no-op if we weren't testing on linux (we were), and we
+also tested calls to `Sys::Platform::IMPL` separately, and that was in the
+clear.  The only time any "work" is done in this method would be the first time
+this method was called, and that work is short circuited once the cached value
+is set.
+
+This left the `require` call.  We updated the script above to include `require
+'config/environment'` and replaced `MiqProcess.processInfo(pid)` with `require
+'linux_admin'` and ran it.
+
+It leaked...
+
+ಠ\_ಠ
+
+
+
+Step 7:  Mind blown
+-------------------
+
+After piecing our heads back together, we noticed a few things:
+
+* Again, without `require 'config/environment'` at the top, no problems,
+  meaning our environment was at play... at least a bit.
+* No matter what we called `require` with, it leaked just the same\*\*\*.
+* We could also recreate this on non-linux operating systems
+
+The second bit should come as no surprise, because after calling require on the
+same file twice, the code isn't reloaded if Ruby determines it has already been
+loaded (hence the difference in return values if you do that in IRB as well:
+`true` for the first time, and `false` every time thereafter).
+
+But still, that means we have a leak in Ruby... right?  But with `require`?
+How has no one run into this before?  Could a c-extension be messing with
+things here or is Ruby the culprit?
+
+At this point, we tried a few different profiling techniques:
+
+* I personally used `gdb` again using a [crazy python gdb script][10] I cobbled
+  together
+* some co-workers used the `Instruments` application to do basically the same
```

Comment:
Minor, but you might want to link to the Instruments app, in case others had never heard of it:

https://developer.apple.com/library/content/documentation/DeveloperTools/Conceptual/InstrumentsUserGuide/index.html

or

https://developer.apple.com/library/content/documentation/Performance/Conceptual/PerformanceOverview/PerformanceTools/PerformanceTools.html

Reply:
Honestly, I didn't know where to find this, so thanks for the links!  (using the first one, by the way)